app/index.ts

```diff
@@ -55,25 +55,98 @@ function createWindow() {
 //
 // Create the window when the application is ready.
 //
-app.on("ready", createWindow);
+app.on("ready", () => {
+    createWindow();
+});
 
 
 //
 // Terminates the application when all windows have been closed.
 //
 app.on("window-all-closed", () => {
-    if (process.platform !== "darwin") {
-        app.quit();
-    }
+    //if (process.platform !== "darwin") {
+    //    app.quit();
+    //}
+
+
+    // It's becoming more common in macOS to just quit the application and 
+    // Save state between launches.
+    app.quit();
 });
 
 
 //
 // Recreates the window if it has been lost while the application was inactive.
 // This will likely occur on mobile devices with limited resources.
 //
-app.on("activate", () => {
-    if (win === null) {
-        createWindow();
+// app.on("activate", () => {
+//     if (win === null) {
+//         createWindow();
+//     }
+// });
+
+
+
+
+
+/** Managing Additional Windows **/
+// TODO: probs should split this into it's own file.
+
+var windows = new Map<Number, [Electron.BrowserWindow, ModalInfo]>();
+
+ipcMain.on('createWindow', (event, arg) => {
+    event.returnValue = createNewWindow(arg);
+});
+
+ipcMain.on('windowResult', (event, arg: ModalInfo) => {
+    if (win) {
+        win.webContents.send('windowResult', arg);
+    }
+
+    var window = windows.get(arg.id);
+    if (window) { // TODO: I think there's syntax sugar to make this more readable. 
+        window[0].close();
     }
 });
+
+ipcMain.on('getWindowInfo', (event, arg: Number) => {
+    var window = windows.get(arg);
+    event.returnValue = window ? window[1] : null;
+});
+
+function createNewWindow(selector: string): Number {
+    if (win) {
+        var newWindow = new BrowserWindow({
+            parent: win,
+            modal: true,
```

Comment:
This supports multiple modal windows (though MacOS only displays one a time.) We can talk about support for multiple root windows but it's not a huge priority for me right now.